libsignal-service/src/cipher.rs

```diff
@@ -83,84 +97,99 @@ impl ServiceCipher {
-                Plaintext {
-                    metadata,
-                    data: Vec::from(buf.as_slice()),
-                }
+                let mut data = cipher
+                    .decrypt_pre_key_message(
+                        &PreKeySignalMessage::deserialize(
+                            &self.context,
+                            ciphertext,
+                        )?,
+                    )?
+                    .as_slice()
+                    .to_vec();
+                let version =
+                    self.store_context.load_session(&sender)?.state().version();
+                strip_padding(version, &mut data)?;
+                Plaintext { metadata, data }
             }
             Type::Ciphertext => {
-                let cipher = SessionCipher::new(
+                let sender = self.get_preferred_protocol_address(
+                    envelope.source_address().expect("Envelope with source"),
+                    envelope.source_device(),
+                )?;
+                let metadata = Metadata {
+                    sender: envelope
+                        .source_address()
+                        .expect("envelope with source"),
+                    sender_device: envelope.source_device() as i32,
+                    timestamp: envelope.timestamp(),
+                    needs_receipt: false,
+                };
+                let mut data = SessionCipher::new(
                     &self.context,
                     &self.store_context,
-                    &address,
-                )?;
-                let buf = cipher.decrypt_message(
-                    &SignalMessage::deserialize(&self.context, ciphertext)?,
-                )?;
-                Plaintext {
-                    metadata,
-                    data: Vec::from(buf.as_slice()),
-                }
+                    &sender,
+                )?
+                .decrypt_message(&SignalMessage::deserialize(
+                    &self.context,
+                    ciphertext,
+                )?)?
+                .as_slice()
+                .to_vec();
+                let version =
+                    self.store_context.load_session(&sender)?.state().version();
+                strip_padding(version, &mut data)?;
+                Plaintext { metadata, data }
             }
             Type::UnidentifiedSender => {
-                unimplemented!("UnidentifiedSender requires SealedSessionCipher, please report a bug against libsignal-service-rs.");
+                let DecryptionResult {
+                    sender_uuid,
+                    sender_e164,
+                    device_id,
+                    padded_message: mut data,
+                    version,
+                } = self
+                    .sealed_session_cipher
+                    .decrypt(ciphertext, envelope.timestamp())?;
+                let sender = ServiceAddress {
+                    e164: sender_e164.unwrap_or("".into()), // XXX: FIXME
```

Comment:
XXX -> this means e164 should become `Option<>`, already? Whisperfish currently always assumes e164, because we're old skool.

Reply:
Yes, I initially wanted to change it. It'll be a slightly annoying breaking change though. Maybe we can defer it for later? (I can open an issue).